src/envoy/http/authn/authenticator_base.h

```diff
@@ -0,0 +1,82 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "authentication/v1alpha1/policy.pb.h"
+#include "common/common/logger.h"
+#include "server/config/network/http_connection_manager.h"
+#include "src/envoy/http/authn/context.pb.h"
+#include "src/envoy/http/authn/filter_context.h"
+
+namespace Envoy {
+namespace Http {
+namespace IstioAuthN {
+
+// AuthenticatorBase is the base class for authenticator. It provides functions
+// to perform individual authentication methods, which can be used to construct
+// compound authentication flow.
+class AuthenticatorBase : public Logger::Loggable<Logger::Id::filter> {
+ public:
+  // Callback type for individual authentication method.
+  typedef std::function<void(const Payload*, bool)> MethodDoneCallback;
+
+  // Callback type for the whole authenticator.
+  typedef std::function<void(bool)> DoneCallback;
+
+  AuthenticatorBase(FilterContext* filter_context,
+                    const DoneCallback& callback);
+  virtual ~AuthenticatorBase();
+
+  // Perform authentication.
+  virtual void run() PURE;
+
+  // Calls done_callback_ with success value.
+  void done(bool success) const;
+
+  // Validates x509 given the params (more or less, just check if x509 exists,
+  // actual validation is not neccessary as it already done when the connection
+  // establish), and extract authenticate attributes (just user/identity for
+  // now). Calling callback with the extracted payload and corresponding status.
+  virtual void validateX509(
+      const istio::authentication::v1alpha1::MutualTls& params,
+      const MethodDoneCallback& done_callback) const;
+
+  // Validates JWT given the jwt params. If JWT is validated, it will call
+  // the callback function with the extracted attributes and claims (JwtPayload)
+  // and status SUCCESS. Otherwise, calling callback with status FAILED.
+  virtual void validateJwt(const istio::authentication::v1alpha1::Jwt& params,
+                           const MethodDoneCallback& done_callback) const;
+
+  // Mutable accessor to filter context.
+  FilterContext* filter_context() { return filter_context_; }
+
+ private:
+  // Pointer to filter state. Do not own.
+  FilterContext* filter_context_;
+
+  const DoneCallback done_callback_;
+};
+
+// Return pointer to credential rule matching with peer_id from the policy.
+// If no rule found, return the default credential rule.
+const istio::authentication::v1alpha1::CredentialRule&
+findCredentialRuleOrDefault(
```

Comment:
why this is a global function?  can it be part of some class?

Reply:
Yes and no. It can run by itself (i.e without any other data), and I didn't know where this should be called (i.e in filter or one of the authenticator). We can move it to the class when it's settle down.

Reply:
how about move it into base class as a static function

Reply:
I'd like to defer this when we know better where this function should be used. Put it in some arbitrary class right now will be as bad.